Hw3

IAGrig opened the pull request.

IAGrig committed:
build: add database module and prettify initial configs

IAGrig committed:
feat: implement bot and all commands, add allowed domains

The bot starts with Spring, gets token from enviromental variables.
Command processing is based on Chain of Responsibility pattern.
The LinksHandler is responsible for processing and validating
incoming links for tracking.
The allowed domains are specified in the configuration files as
a single line separated by commas without spaces.

IAGrig committed:
feat: add database support

The Stub Database is a temporal in-memory database, which will be
replaced by a real database in the future.

IAGrig committed:
test: add tests for commands handling and LinksHandler

IAGrig committed:
feat: add a menu with commands to the bot and do a little refactoring

IAGrig committed:
fix: fix a bug in /list (links id were not displayed)

The user must to specify the link id to stop tracking the link using
/untrack, but these ids were not displayed in /list output.

IAGrig committed:
fix: correct all defects

The application configuration has been moved from the dev profile
to the general one. Unused fields in classes have been removed.
Fixed the processing of the /start command and the processing of
allowed domains, added functionality for tracking registered users.

IAGrig committed:
test: fix the tests - replace junit4 with junit5

IAGrig committed:
Merge branch 'hw1' into hw3

IAGrig committed:
feat(bot): add a method for sending messages to the user

IAGrig committed:
feat(database): add a method for deleting users and links

IAGrig committed:
feat(bot): add API controller, DTOs, service, exception handler

IAGrig committed:
build(scrapper): add database dependency to scrapper

IAGrig committed:
feat(scrapper): add API controller, DTOs, services, exception handler

IAGrig committed:
refactor: tiny refactoring

IAGrig committed:
build: add swagger-ui

IAGrig committed:
refactor: separate DTOs into the models module

IAGrig committed:
refactor: move http clients to a separate package

IAGrig committed:
feat(bot): add http client for scrapper

IAGrig committed:
feat(scrapper): add http client for bot

IAGrig committed:
fix(models): make DTOs to use URIs for links

IAGrig committed:
fix: make track/untrack endpoints return LinkResponse

IAGrig committed:
feat: add advanced exception handling

damir2407 commented on bot/src/main/java/edu/java/bot/controllers/BotController.java:
```diff
@@ -0,0 +1,22 @@
+public class BotController {
+    private UpdateService updateService;
+
+    @PostMapping(path = "/updates")
```
по спеке /updates должен быть на уровне контроллера

damir2407 commented on models/src/main/java/edu/java/dto/LinkUpdateRequest.java:
```diff
@@ -0,0 +1,63 @@
+@Data
+@Validated
+public class LinkUpdateRequest {
+    @JsonProperty("id")
```
Нет нужны в использовании JsonProperty с именами равными именам модели. Спринг все прекрасно спарсит и так

damir2407 commented on bot/src/main/java/edu/java/bot/httpClients/ScrapperHttpClient.java:
```diff
@@ -0,0 +1,99 @@
+import reactor.core.publisher.Mono;
+
+@AllArgsConstructor
+public class ScrapperHttpClient {
```
Зарегистрируй его как бин также как и клиенты гитхаба и стаковерфлоу

damir2407 commented on bot/src/main/java/edu/java/bot/httpClients/ScrapperHttpClient.java:
```diff
@@ -0,0 +1,99 @@
+@AllArgsConstructor
+public class ScrapperHttpClient {
+    private final WebClient client;
+    private final String baseUrl;
```
и где юзаешь? Сделай также как во второй домашке через проперти

damir2407 commented on models/src/main/java/edu/java/dto/ApiErrorResponse.java:
```diff
@@ -0,0 +1,62 @@
+@Validated
+@Data
+public class ApiErrorResponse {
+    @JsonProperty("description")
```
Пройдись плиз по всем моделям и почисти JsonProperty

damir2407 commented on scrapper/src/main/java/edu/java/controllers/ScrapperController.java:
```diff
@@ -0,0 +1,64 @@
+    private UserService userService;
+    private LinkService linkService;
+
+    @PostMapping(path = "/tg-chat/{id}")
```
/tg-chat должен быть на уровне контроллера, посмотри внимательнее спеку

damir2407 commented on scrapper/src/main/java/edu/java/controllers/ScrapperController.java:
```diff
@@ -0,0 +1,64 @@
+        return ResponseEntity.ok("Чат успешно удалён");
+    }
+
+    @GetMapping(path = "/links")
```
/links тоже должен быть на уровне контроллера, то-есть тебе нужно создать отдельный контроллер под /links операции

damir2407 commented on scrapper/src/main/java/edu/java/httpClients/BotHttpClient.java:
```diff
@@ -0,0 +1,32 @@
+import reactor.core.publisher.Mono;
+
+@AllArgsConstructor
+public class BotHttpClient {
```
Зарегай также как бин вместе с гитхабом и стаковерфлоу с возможностью указать базовый урл как и у них

IAGrig committed:
fix: remove useless @JsonProperty, move endpoints to controllers level

Move /updates to the controller level in UpdatesController
Move /tg-chat to the controller level in TgChatConroller
Move /links to the controller level in LinksController

IAGrig committed:
fix: register http clients as beans, add configuration for them

IAGrig committed:
test: add jdbc dependencies

IAGrig committed:
feat: add liquibase migrations

IAGrig committed:
test: add the testcontainer integration test

IAGrig committed:
fix: fix a bug in ScrapperHttpClient, fix URI validation

IAGrig commented:
@damir2407 У меня реализован свой UriValidator. Есть мысль засунуть в него проверку на допустимый домен (github, stackoverflow). Но allowed-domains у меня уже перечислены в конфиге бота. Дублировать точно не хочется, потом легко забыть поменять в двух местах.
Как лучше получить доступ к этим allowed-domains. Создать свой конфиг для models и делать UriValidator бином? При этом убрать allowed-domains из бота, чтобы вся валидация была в модуле models?

IAGrig committed:
Merge branch 'hw4' into hw5, move database tests to separate package

IAGrig committed:
refactor: move command handlers to stub package, add User, updade Link

IAGrig committed:
feat: add JDBC repositories for links and users

IAGrig committed:
test: add test for JDBC link and user repositories

IAGrig committed:
feat: add JDBC services, rewrite controllers, add command handlers

IAGrig committed:
feat: implement scheduled changes checker, connect scrapper and bot

IAGrig committed:
fix: fix URI validation

IAGrig committed:
feat: add a detailed update notifications for Stackoverflow

IAGrig committed:
feat: make my plain java entities real persistence entities

IAGrig committed:
feat: add JPA repositories and services

IAGrig committed:
test: add JPA services and repositories tests

IAGrig committed:
feat: implement a configurable database backend selection

IAGrig committed:
fix: add a test configuration for JPA services tests, fix repo tests

IAGrig committed:
refactor: separate link checkers to beans

IAGrig committed:
refactor: make simple beans components

IAGrig committed:
feat: add retry managers to all HTTP clients

IAGrig committed:
feat: add retry codes and linear retry

IAGrig committed:
feat(bot): add bucket4j rate limit for public endpoints

IAGrig committed:
feat(scrapper): add bucket4j rate limit for public endpoints

IAGrig committed:
fix: add missed and remove useless annotations, add Optionals to JPA

IAGrig committed:
feat: add kafka and zookeeper to the compose.yml

IAGrig committed:
feat(scrapper): add kafka properties, producers and updates handler

IAGrig committed:
feat(bot): add kafka properties and listener service

IAGrig committed:
fix(scrapper): add missing annotations, fix conditional bean creation

IAGrig committed:
fix(bot): fix configs, add updates handling and missing annotation

IAGrig committed:
feat(bot): implement Dead Letter Queue

IAGrig committed:
fix(scrapper): add null check in update method

IAGrig committed:
test(scrapper): add tests to the scrapper kafka producer

IAGrig committed:
feat(bot): add metrics and micrometer dependencies

IAGrig committed:
feat(scrapper): add metrics and micrometer dependencies

IAGrig committed:
feat: add prometheus config, grafana and prometheus images to the compose

IAGrig committed:
feat(bot): add a custom metric - count of processed updates

IAGrig committed:
feat(scrapper): add a custom metric - count of processed updates

IAGrig committed:
fix(bot): change bot handling of null tgIds, remove a partial sending

IAGrig committed:
fix: fix broken tests, add correct tests to bot kafka consumer

IAGrig committed:
fix(scrapper): fix the processing of adding already tracked links

IAGrig committed:
test(scrapper): fix problem with same user IDs in parallel tests

IAGrig committed:
feat(bot): GitHub actions for publishing to the GitHub Container Registry

IAGrig committed:
feat(scrapper): GitHub actions for publishing to the GitHub Registry

IAGrig committed:
feat(scrapper): add JOOQ to the configuration

IAGrig committed:
feat(scrapper): add JOOQ codegen and generated classes

IAGrig committed:
feat(scrapper): implement JOOQ repositories and services

IAGrig committed:
test(scrapper): add tests for JOOQ services and repositories

IAGrig committed:
Merge branch 'hw5-bonus-jooq' into hw10

IAGrig committed:
Merge branch 'hw10' into hw9

IAGrig committed:
Merge branch 'hw9' into hw8

IAGrig committed:
Merge branch 'hw8' into hw7

IAGrig committed:
Merge branch 'hw7' into hw6

IAGrig committed:
Merge branch 'hw6' into hw5-bonus

IAGrig committed:
fix(scrapper): remove unnecessary @Qualifier annotations

IAGrig committed:
Merge branch 'hw5-bonus' into hw5

IAGrig committed:
Merge branch 'hw5' into hw3
